helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/IZkStateListener.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2010 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.helix.manager.zk.zookeeper;
+
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+
+public interface IZkStateListener {
+
+  /**
+   * Called when the zookeeper connection state has changed.
+   *
+   * @param state
+   *            The new state.
+   * @throws Exception
+   *             On any error.
+   */
+  void handleStateChanged(KeeperState state) throws Exception;
+
+  /**
+   * Called after the zookeeper session has expired and a new session has been created.
+   * You would have to re-create any ephemeral nodes here.
+   *
+   * @deprecated
+   * This is deprecated, because it may cause session race condition when creating ephemeral nodes.
+   * Please use {@link #handleNewSession(String)} instead, which takes care of race condition.
+   *
+   * @throws Exception
+   *             On any error.
+   */
+  @Deprecated
+  void handleNewSession() throws Exception;
```

Comment:
As a new interface, can we just not add this method that is only for deprecating?

The benefit of having this here was for, if the new interface is inherited from the IOItech interface, we ensure strong backward compatibility. But now that we are separating them cleanly, the user who implements their listener based on IOItech interface will still work with the HelixZkClient. I think there is no need to keep this Deprecated method in the new interface.

Reply:
I think not adding this interface has the benefit of "forcing" the user to use `handleNewSession(sessionId)`. Isn't that what we want to achieve?

Another way to put this is, do we ever have cases where a user wouldn't care whether `sessionIds` match or not? (I can't think of any - people would assume that sessionIds match?).

Reply:
From another perspective, `handleNewSession()` could bring in a bug or race condition. If we fix the bug, we may just remove it and force users to use `handleNewSession(sessionId)`. I think this consideration makes more sense to me. Let me get @lei-xia also to discuss about this.